File: src/main/java/com/example/librarymanagement/dto/mapper/UserMapper.java
package com.example.librarymanagement.dto.mapper;

import com.example.librarymanagement.dto.UserRequestDto;
import com.example.librarymanagement.dto.UserResponseDto;
import com.example.librarymanagement.model.entity.User;
import org.mindrot.jbcrypt.BCrypt;
import org.springframework.stereotype.Component;

import java.util.ArrayList;
import java.util.List;

@Component
public class UserMapper {
    private static final String salt = BCrypt.gensalt(12);

    public User toUser(UserRequestDto userRequestDto) {
        User user = new User();
        user.setEmail(userRequestDto.getEmail());
        user.setFirstName(userRequestDto.getFirstName());
        user.setLastName(userRequestDto.getLastName());
        user.setPhoneNumber(userRequestDto.getPhoneNumber());
        user.setPassword(hashPassword(userRequestDto.getPassword()));
        return user;
    }

    public UserResponseDto toUserResponseDto(User user) {
        UserResponseDto userResponseDto = new UserResponseDto();
        userResponseDto.setId(user.getId());
        userResponseDto.setEmail(user.getEmail());
        userResponseDto.setFirstName(user.getFirstName());
        userResponseDto.setLastName(user.getLastName());
        userResponseDto.setPhoneNumber(user.getPhoneNumber());
        return userResponseDto;
    }

    public List<UserResponseDto> toUserResponseDto(List<User> usersList) {
        if(usersList == null || usersList.isEmpty()) return new ArrayList<>();

        return usersList.stream()
                .map(this::toUserResponseDto)
                .toList();
    }

    private String hashPassword(String plainPassword) {
        return BCrypt.hashpw(plainPassword, salt);
    }
}

File: src/main/java/com/example/librarymanagement/dto/mapper/UserMapper.kt
package com.example.librarymanagement.dto.mapper

import com.example.librarymanagement.dto.UserRequestDto
import com.example.librarymanagement.dto.UserResponseDto
import com.example.librarymanagement.model.domain.User
import org.mindrot.jbcrypt.BCrypt
import org.springframework.stereotype.Component

@Component
class UserMapper {
    private val salt = BCrypt.gensalt()

    fun toUser(userRequestDto: UserRequestDto, id: String? = null) : User {
        return User(
            id = id,
            firstName = userRequestDto.firstName,
            lastName = userRequestDto.lastName,
            email = userRequestDto.email,
            password = hashPassword(userRequestDto.password),
            phoneNumber = userRequestDto.phoneNumber
        )
    }

    fun toUserResponseDto(user: User) : UserResponseDto {
        return UserResponseDto(
            id = user.id!!,
            firstName = user.firstName,
            lastName = user.lastName,
            email = user.email,
            phoneNumber = user.phoneNumber
        )
    }

    fun toUserResponseDto(user: List<User>) : List<UserResponseDto> = user.map { toUserResponseDto(it) }

    fun hashPassword(password: String): String = BCrypt.hashpw(password, salt)
}
